core/src/connection_reuse.rs

```diff
@@ -40,193 +40,543 @@
+    transport: UpgradedNode<T, C>,
+
+    /// All the connections that were opened, whether successful and/or active or not.
+    // TODO: this will grow forever
+    connections: Mutex<FnvHashMap<Multiaddr, PeerState<C::Output>>>,
+
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<HashMap<usize, task::Task, UsizeHasherBuilder>>,
 
-    // Struct shared between most of the `ConnectionReuse` infrastructure.
-    shared: Arc<Mutex<Shared<C::Output>>>,
+    /// Counter giving us the next listener_id
+    listener_counter: Counter,
 }
 
-struct Shared<M>
+impl<T, C> ConnectionsManager<T, C>
 where
-    M: StreamMuxer,
+    T: Transport,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
+    C::Output: StreamMuxer + Clone,
 {
-    // List of active muxers.
-    active_connections: FnvHashMap<Multiaddr, M>,
+    /// Notify all `Tasks` in `self.notify_on_new_connec` that a new connection was established
+    /// consume the tasks in that process
+    fn new_con_notify(&self) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        trace!("Notify {} listeners about a new connection", notifiers.len());
+        for (_, task) in notifiers.drain() {
+            task.notify();
+        }
+    }
 
-    // List of pending inbound substreams from dialed nodes.
-    // Only add to this list elements received through `add_to_next_rx`.
-    next_incoming: Vec<(M, M::InboundSubstream, Multiaddr)>,
+    /// Register `task` under id if not yet registered
+    fn register_notifier(&self, id: usize, task: task::Task) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        notifiers.insert(id, task);
+    }
 
-    // New elements are not directly added to `next_incoming`. Instead they are sent to this
-    // channel. This is done so that we can wake up tasks whenever a new element is added.
-    add_to_next_rx: mpsc::UnboundedReceiver<(M, M::InboundSubstream, Multiaddr)>,
+    /// Clear the cached connection if the entry contains an error. Returns whether an error was
+    /// found and has been removed.
+    fn clear_error(&self, addr: Multiaddr) -> bool {
+        let mut conns = self.connections.lock();
+
+        if let Entry::Occupied(e) = conns.entry(addr) {
+            if let PeerState::Errored(ref err) = e.get() {
+                trace!(
+                    "Clearing existing connection to {} which errored earlier: {:?}",
+                    e.key(),
+                    err
+                );
+            } else {
+                return false; // nothing to do, quit
+            }
 
-    // Other side of `add_to_next_rx`.
-    add_to_next_tx: mpsc::UnboundedSender<(M, M::InboundSubstream, Multiaddr)>,
+            // clear the error
+            e.remove();
+            true
+        } else {
+            false
+        }
+    }
+}
+
+fn poll_for_substream<M>(mut outbound: M::OutboundSubstream, addr: &Multiaddr)
```

Comment:
What does this function do?
I don't care if a function doesn't have documentation if it is trivial, but this function clearly is not.